### PR TITLE
teams now correctly written to json file

### LINKED ISSUE #69: wrong content written to files
When the program writes to json files (settings, teams, scores etc.), the resulting file is just:

```json
[object Object]
```

### PR DESCRIPTION
**Pull request by itamargreen**
this one fixes #69 . Only for teams for now... 

the solution was to add a "post" handler for the teams view router. when saving, an http post message is sent and is handled, server-side, by the teams view router.

### CODE REVIEW

**Comment by itamargreen** [comment]
It's failing the checks because stuff was added to the expected behavior, and so the checks need to be updated.

I'm not sure how to tell it to expect http post calls.

File: src/js/services/ng-settings.js [hunk 1]
```diff
@@ -55,6 +55,12 @@ define('services/ng-settings',[
                 };
                 //create settings file if not there
                 log('settings read error, trying to create file', err);
+                var data = { settings: defaults };
+                $http.post("/settings/save", data).success(function (data, status) {
+                    log('Data posted successfully');
+                }).error(function () {
+                    log('failed retrieving settings');
+                });
                 return $fs.write('settings.json',defaults).then(function() {
```
Comment (idanstark42) [inline src/js/services/ng-settings.js]: You're saving here with http and with fs? Please use only one way...

File: server_modules/data.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
+var utils = require('./utils');
+var fileSystem = require('./file_system');
+var authorize = require('./auth').authorize;
+var reduceToMap = require('./scores').reduceToMap;
```
Comment (idanstark42) [inline server_modules/data.js]: If this funciton is used in more then one module, why isn't it in utils? that's where it belongs

File: src/js/services/ng-scores.js [hunk 3]
```diff
@@ -18,7 +18,7 @@ define('services/ng-scores',[
 
     // Current file version for scores.
     // Increment when adding/removing 'features' to stored scores.
-    var SCORES_VERSION = 3;
+    var SCORES_VERSION = 2;
```
Comment (idanstark42) [inline src/js/services/ng-scores.js]: ?

Comment (itamargreen) [reply src/js/services/ng-scores.js]: this doesn't belong here....

File: server_modules/data.js (hunk 2 above)
Comment (itamargreen) [reply server_modules/data.js]: so, just a module that detects a `post` that ends with `/save/` and then it does the save? for the `scores` part, that might not work...

Comment (idanstark42) [reply server_modules/data.js]: No. Something like this
// file.js
exports.route = function(app, filename) {
  app.get(`/${filename}`,function(req,res) {
        fileSystem.readJsonFile(fileSystem.getDataFilePath(`${filename}.json`)).then(function(result) {
            res.json(result);
        }).catch(err => utils.sendError(res, err)).done();
    });

    app.post(`/${filename}/save`, function (req, res, next) {
         var stages = JSON.stringify(req.body.stages);
         fileSystem.writeFile(fileSystem.getDataFilePath(`${filename}.json`), stages).then(function () {
             res.status(200).end();
         }).catch(function (err) {
             utils.sendError(res, { status: 500, message: `error writing file: ${filename}` })
         });
     });
}

// stages.js
var file = require('./file');
exports.route = function(app) {
  file.route(app, 'stages');
}

Comment (itamargreen) [reply server_modules/data.js]: @idanstark42 I'm not sure I understand this...

Comment (idanstark42) [reply server_modules/data.js]: Ok. What we have here are a few files. One of them is a utility file, and the others are and actual routing modules.

file.js exposes one function. This function recieves two parameters: `app` and `filename`. `app` is a node.js application, and `filename` is a string. This function creates routes for the app to acess a file with the givven filename in the data folder, for both read and write.

The stages.js module uses that function to create routes for stages.json. The teams.js module uses that function to create routes for teams.json, and so forth.

Comment (itamargreen) [reply server_modules/data.js]: the main problem is that `$fs.write` returns a promise and I'm not sure how to replace that with an http post.

Comment (itamargreen) [reply server_modules/data.js]: I'm getting err_connection_refused...

Comment (idanstark42) [reply server_modules/data.js]: You shuold read a bit about promises. There are simple ways to create them on your own

Comment (itamargreen) [reply server_modules/data.js]: @idanstark42 apparently `$http.post` and `.get` return promises...

File: src/js/services/ng-settings.js [hunk 4]
```diff
@@ -8,7 +8,7 @@ define('services/ng-settings',[
 ],function(module,log) {
     "use strict";
 
-    return module.service('$settings', ["$fs", function($fs) {
+    return module.service('$settings', ["$fs","$http", function($fs,$http) {
```
Comment (idanstark42) [inline src/js/services/ng-settings.js]: You can remove the fs dependency

File: src/js/services/ng-teams.js [hunk 5]
```diff
@@ -8,7 +8,7 @@ define('services/ng-teams',[
 ],function(module,log) {
     "use strict";
 
-    return module.service('$teams', ["$fs", function($fs) {
+    return module.service('$teams', ["$fs","$http", function($fs,$http) {
```
Comment (idanstark42) [inline src/js/services/ng-teams.js]: You can remove the fs dependency

File: src/js/views/settings.js [hunk 6]
```diff
@@ -46,6 +46,12 @@ define('views/settings',[
             };
 
             $scope.save = function() {
+                var data = { settings: $settings.settings };
+                $http.post("/settings/save", data).success(function (data, status) {
```
Comment (idanstark42) [inline src/js/views/settings.js]: It seems to me like settings.save already does that

File: src/js/views/settings.js [hunk 7]
```diff
@@ -56,6 +62,12 @@ define('views/settings',[
                 stages.forEach(function(stage) {
                     return $stages.add(stage);
                 });
+                var data = { stages: $stages._rawStages };
+                $http.post("/stages/save", data).success(function (data, status) {
```
Comment (idanstark42) [inline src/js/views/settings.js]: It seems to me that stages.save already does that.

File: server_modules/files.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+        }).catch(err => utils.sendError(res, err)).done();
+    });
+
+    app.post(`/${filename}/save`,authorize.any, function (req, res, next) {
+
+        var data = JSON.stringify(req.body[`${filename}`]);
+        fileSystem.writeFile(fileSystem.getDataFilePath(`${filename}.json`), data).then(function () {
+            res.status(200).end();
+        }).catch(function (err) {
+            utils.sendError(res, { status: 500, message: `error writing file: ${filename}` })
```
Comment (idanstark42) [inline server_modules/files.js]: After the last change I did in the logs PR, the sendError method changed to:
`res.sendError({ status: 404, message: 'file not found ${file}' });`

File: server_modules/files.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+var fileSystem = require('./file_system');
+var log = require('./log').log;
+var authorize = require('./auth').authorize;
+
+exports.route = function (app, filename) {
+
+    app.get(`/${filename}`, function (req, res) {
+        fileSystem.readJsonFile(fileSystem.getDataFilePath(`${filename}.json`)).then(function (result) {
+            res.json(result);
+        }).catch(err => utils.sendError(res, err)).done();
```
Comment (idanstark42) [inline server_modules/files.js]: After the last change I did in the logs PR, the sendError method changed to:
res.sendError({ status: 404, message: 'file not found ${file}' });

File: src/js/services/ng-stages.js [hunk 10]
```diff
@@ -51,19 +51,19 @@ define('services/ng-stages',[
...
-        Stages.prototype.save = function() {
-            return $fs.write('stages.json', this._rawStages).then(function() {
-                log('stages saved');
-            }, function(err) {
-                log('stages write error', err);
+        Stages.prototype.save = function () {
+            return $http.post("/stages/save", { stages: this.stages }).success(function (data, status) {
+                log('Data posted successfully');
+            }).error(function () {
+                log('failed retrieving stages');
```
Comment (idanstark42) [inline src/js/services/ng-stages.js]: You don't retrieve the stages here, you're saving it

File: spec/services/ng-settingsSpec.js [hunk 11]
```diff
@@ -57,16 +75,17 @@ describe('ng-settings',function() {
                 currentStage: 'practice',
                 ignoreNegativeScores: true
             }
-            fsMock.read.and.returnValue($q.reject('no file'));
-            fsMock.write.and.returnValue($q.when());
             $settings.load();
             $rootScope.$digest();
-            expect(fsMock.write).toHaveBeenCalledWith('settings.json',defaults);
             expect($settings.settings).toEqual(defaults);
```
Comment (idanstark42) [inline spec/services/ng-settingsSpec.js]: This shuold be called in a `then` of the `$settings.load` function

File: spec/services/ng-stagesSpec.js [hunk 12]
```diff
@@ -42,18 +48,25 @@ describe('ng-stages',function() {
 
     describe('init',function() {
         it('should load stages by default', function() {
+            return $stages.init().then(function(){
```
Comment (idanstark42) [inline spec/services/ng-stagesSpec.js]: The double call to `$stages.init` causes problems here. Call it individually on each function.

Comment (itamargreen) [reply spec/services/ng-stagesSpec.js]: That explains so much

**Comment by itamargreen** [comment]
What about the parts with the http requests? Did i do those correctly?

**Comment by itamargreen** [comment]
@idanstark42 I'm having trouble with one specific test:

```js
it('should log an error if reading fails',function() {
    httpMock.post.and.returnValue(Q.reject('aargh'));
    return $stages.save().then(function() {
        expect(logMock).toHaveBeenCalledWith('stages write error','aargh');
    });
});
```

For some reason, this happens:


`Expected spy log to have been called with [ 'stages write error', 'aargh' ] but actual calls were [ 'after OK: undefined' ], [ 'Indexing none-array: foo' ], [ 'Indexing none-array: 42' ], [ 'Indexing none-array: undefined' ], [ 'fail getting file system ' ], [ 'fail getting path fllscoring' ], [ 'fail getting directory fllscoring' ], [ 'fail getting file system ' ], [ 'fail getting path fllscoring' ], [ 'fail getting file system ' ], [ 'fail getting path fllscoring' ], [ 'fail getting writer fllscoring/foo.txt' ], [ 'challenge 'foo' loaded' ], [ 'scores loaded, version 3' ], [ 'scores loaded, version 3' ], [ 'scores read error', 'read err' ], [ 'scores loaded, version 3' ], [ 'scores loaded, version 3' ], [ 'auto-broadcasting stage practice' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ], [ 'scores read error', 'read err' ].`

which means that for some reason all the `log(...)` calls through out all the tests are not being cleared.

**Pull request closed without merging**